### PR TITLE
HPCC-16906 Add a new plugin for embedded Python3 code

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:

Cloned the python2 regression suite tests, and ran them

### CODE REVIEW

**Comment by AttilaVamos** [comment]
@richardkchapman It was Python3 environment problem, fixed. This PR rescheduled.

File: plugins/py3embed/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            ./../../roxie/roxiemem
+            ./../../system/jlib)
+
+        add_definitions(-D_USRDLL -DPY3EMBED_EXPORTS)
+
+        HPCC_ADD_LIBRARY(py3embed SHARED ${SRCS})
+        if(${CMAKE_VERSION} VERSION_LESS "2.8.9")
+            message(WARNING "Cannot set NO_SONAME. shlibdeps will give warnings when package is installed")
+        elseif(NOT APPLE)
+            set_target_properties(pyembed PROPERTIES NO_SONAME 1)
```
Comment (AttilaVamos) [inline plugins/py3embed/CMakeLists.txt]: I think this should be 
```
set_target_properties(py3embed PROPERTIES NO_SONAME 1)
```

Comment (richardkchapman) [inline plugins/py3embed/CMakeLists.txt]: I think you are right...

Comment (richardkchapman) [inline plugins/py3embed/CMakeLists.txt]: repushed to fix

**Comment by AttilaVamos** [comment]
@richardkchapman It seems this result is much better. I will tune a bit the log processing and GitHub comment generation and reschedule it.

**Comment by richardkchapman** [comment]
@ghalliday Please review.

There's a lot of code in common with the Python2 plugin, and I slightly wonder if I should common them up - but then again I don't tend to like having one .cpp file compiled more than once (and they can't share the obj files - the python headers they include are different...

**Review by ghalliday — DISMISSED** [review]
@richardkchapman a few minor comments/discussion items.

File: common/dllserver/thorplugin.cpp [hunk 2]
```diff
@@ -634,7 +634,7 @@ bool SafePluginMap::addPlugin(const char *path, const char *dllname)
         if (!dll)
         {
             Owned<PluginDll> n = new PluginDll(path, NULL);
-            if (!n->load(true, false) || !n->init(pluginCtx))
+            if (!n->load(false, false) || !n->init(pluginCtx))
```
Comment (richardkchapman) [inline common/dllserver/thorplugin.cpp]: There would be potential implications if we had multiple plugins that shared a common library that was not linked in by the system - might end up with two copies of it - but I'm not aware that we do so anywhere. Regression suite all seems fine.

I can put in a comment explaining the decision.

File: plugins/py3embed/py3embed.cpp [hunk 3]
```diff
@@ -0,0 +1,1720 @@
...
+    PyObject* err = PyErr_Occurred();
+    if (err)
+    {
+        OwnedPyObject pType, pValue, pTraceBack;
+        PyErr_Fetch(pType.ref(), pValue.ref(), pTraceBack.ref());
+        OwnedPyObject valStr = PyObject_Str(pValue);
+        PyErr_Clear();
+        assertex(PyUnicode_Check(valStr));
+        OwnedPyObject temp_bytes = PyUnicode_AsEncodedString(valStr, ASCII_LIKE_CODEPAGE, "ignore");
+        failx("%s", PyBytes_AsString(temp_bytes));
```
Comment (richardkchapman) [inline plugins/py3embed/py3embed.cpp]: Yes, I wasn't sure whether ascii or utf8 was more appropriate here though in practice I doubt it makes a lot of difference. I'll change to utf8 if you prefer?

File: plugins/py3embed/py3embed.cpp [hunk 4]
```diff
@@ -0,0 +1,1720 @@
...
+#include "rtlfield_imp.hpp"
+#include "nbcd.hpp"
+#include "roxiemem.hpp"
+#include "enginecontext.hpp"
+
+#if PY_MAJOR_VERSION >=3
+  #define Py_TPFLAGS_HAVE_ITER 0
+#endif
+
+#define ASCII_LIKE_CODEPAGE "iso-8859-1"
```
Comment (richardkchapman) [inline plugins/py3embed/py3embed.cpp]: I will change

**Comment by richardkchapman** [comment]
@ghalliday Plese re-review/merge

**Review by ghalliday — DISMISSED** [review]
@richardkchapman 
The following examples in the LN regression suite fail:

bug49457.eclxml
rnaren2.eclxml

Probably because a tree is transformed inconsistently.

File: plugins/py3embed/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+project(py3embed)
+if(PY3EMBED)
+    unset(PYTHONLIBS_FOUND CACHE)
+    unset(PYTHON_LIBRARY CACHE)
+    unset(PYTHON_LIBRARIES CACHE)
+    unset(PYTHON_INCLUDE_PATH CACHE)
+    unset(PYTHON_INCLUDE_DIR CACHE)
+    unset(PYTHON_DEBUG_LIBRARIES CACHE)
+    unset(PYTHONLIBS_VERSION_STRING CACHE)
+    ADD_PLUGIN(py3embed PACKAGES PythonLibs MINVERSION 3.2)
```
Comment (ghalliday) [inline plugins/py3embed/CMakeLists.txt]: PyUnicode_AsUTF8AndSize() is only added in 3.3, so need to increase this version.

**Review by ghalliday — DISMISSED** [review]
@richardkchapman one comment

**Comment by richardkchapman** [comment]
@ghalliday repushed. Let me know when you are done reviewing and I can squash

**Comment by ghalliday** [comment]
@richardkchapman please can you squash

**Comment by richardkchapman** [comment]
@ghalliday Squashed

**Pull request merged**